feat: Add Sphinx Documentation

coolalexzb opened the pull request:
This PR aims to add Sphinx documentation for `pyhf-benchmark`. 
The main page of documentation has been completed, and the page looks like as follows:


<img width="1265" alt="截屏2020-07-27 下午10 24 32" src="https://user-images.githubusercontent.com/19755247/88615658-05217680-d058-11ea-9421-a42f597972db.png">

More contents need to be added into the introduction, examples, etc.

matthewfeickert commented on docs/source/conf.py:
```diff
@@ -0,0 +1,56 @@
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+# ones.
+extensions = ["m2r2"]
```
This is fine for now, but given that [`m2r2`](https://github.com/CrossNox/m2r2#m2r-the-original) is a fork of an abandoned project it is probably worth just switching to RST for long term support, which `pyhf` reluctantly did. That can be for another PR though. :+1:

coolalexzb replied:
Okay, since I wanted to use README.md directly rather than write README.rst before. I will discard the usage of m2r2 and wrote a README.rst file in the next PR.

matthewfeickert replied:
Yeah keeping it as Markdown I think is fine for now.

matthewfeickert commented:
For Sphinx to really be powerful it will need to be able to understand the code though, so it will help to get PR #19 done first.

coolalexzb commented:
@matthewfeickert Documentation contents have been added and you can enter into `docs` directory and use `sphinx-build -b html source build` command to look through the documentation contents.
<img width="1086" alt="截屏2020-08-14 下午4 59 58" src="https://user-images.githubusercontent.com/19755247/90295873-ae85ac00-de4f-11ea-9be1-015de339af5b.png">
<img width="1088" alt="截屏2020-08-14 下午4 59 54" src="https://user-images.githubusercontent.com/19755247/90295878-b1809c80-de4f-11ea-948d-ff512cb34bed.png">

coolalexzb commented:
@matthewfeickert 
One question: In example pages(example.rst), I want to load jupyter notebook contents but when I fail to use the command `pyhf-benchmark run --help ` in the `ipynb` file and the error information is following:
<img width="1120" alt="截屏2020-08-14 下午4 46 16" src="https://user-images.githubusercontent.com/19755247/90295752-61a1d580-de4f-11ea-8c0f-f826af8b80bd.png">

Can you help me to solve the problem?

coolalexzb commented:
@matthewfeickert  What are the further steps for this PR?

matthewfeickert commented on docs/source/conf.py:
```diff
@@ -0,0 +1,100 @@
+xref_links = {"arXiv:1007.1727": ("[1007.1727]", "https://arxiv.org/abs/1007.1727")}
+
+# Github repo
+issues_github_path = "scikit-hep/pyhf/pyhf-benchmark"
```
```suggestion
issues_github_path = "pyhf/pyhf-benchmark"
```

path is wrong

matthewfeickert commented on docs/source/development.rst:
```diff
@@ -0,0 +1,58 @@
+
+    pre-commit install
+
+Testing
+-------
+
+TestPyPI
+~~~~~~~~
+
+``pyhf-benchmark`` tests packaging and distributing by publishing each commit to
+``master`` to `TestPyPI <https://test.pypi.org/project/pyhf-benchmark/>`__.
+In addition, installation of the latest test release from TestPyPI can be tested
+with
+
+.. code-block:: bash
+
+  python -m pip install --extra-index-url https://test.pypi.org/simple/ --pre pyhf-benchmark
+
+.. note::
+
+  This adds TestPyPI as `an additional package index to search <https://pip.pypa.io/en/stable/reference/pip_install/#cmdoption-extra-index-url>`__
+  when installing ``pyhf-benchmark`` specifically.
+  PyPI will still be the default package index ``pip`` will attempt to install
+  from for all dependencies.
+
+Publishing
+----------
+
+Publishing to `PyPI <https://pypi.org/project/pyhf-benchmark/>`__ and `TestPyPI <https://test.pypi.org/project/pyhf-benchmark/>`__
+is automated through the `PyPA's PyPI publish GitHub Action <https://github.com/pypa/gh-action-pypi-publish>`__
+and the ``pyhf-benchmark`` `Tag Creator GitHub Actions workflow <https://github.com/scikit-hep/pyhf-benchmark/blob/master/.github/workflows/tag.yml>`__.
+A release can be created from any PR created by a core developer by adding a
+``bumpversion`` tag to it that corresponds to the release type:
+`major <https://github.com/scikit-hep/pyhf-benchmark/labels/bumpversion%2Fmajor>`__,
+`minor <https://github.com/scikit-hep/pyhf-benchmark/labels/bumpversion%2Fminor>`__,
+`patch <https://github.com/scikit-hep/pyhf-benchmark/labels/bumpversion%2Fpatch>`__.
+Once the PR is tagged with the label, the GitHub Actions bot will post a comment
+with information on the actions it will take once the PR is merged. When the PR
+has been reviewed, approved, and merged, the Tag Creator workflow will automatically
+create a new release with ``bumpversion`` and then deploy the release to PyPI.
```
At the moment none of this makes sense to include as this isn't done. In general please don't copy and paste things from pyhf. It is better to look at what pyhf does but then write things for yourself.
```suggestion
```

matthewfeickert commented on src/pyhf_benchmark/manager.py:
```diff
@@ -15,16 +21,31 @@ def __init__(self, meta=None):
+
+        Args:
+            meta: Meta information for a run
+
+        Returns:
+
```
```suggestion
```

If no return don't list it.

matthewfeickert commented on src/pyhf_benchmark/plot.py:
```diff
@@ -98,6 +127,20 @@ def subplot(y_label, column, output, directory, filename):
+        directory: Directory for plot file
+        filename: Filename for plot file
+
+    Returns:
+
```
As I mentioned before, probably makes more sense to have the function accept an axis object and then return an axis object then try to wrap all of it, but that can be done for a later PR. At the moment this doesn't return anything so it should get fixed.

matthewfeickert commented on src/pyhf_benchmark/plot.py:
```diff
@@ -98,6 +127,20 @@ def subplot(y_label, column, output, directory, filename):
+
+    """
     fig, ax = plt.subplots()
     ax.set_xlabel("Time (minutes)")
```
This should probably be seconds not minutes.

matthewfeickert commented:
None of the required dependencies are setup in a `docs` extra in `setup.py` so there is no way to build the docs at the moment if you were a developer who just installed everything. All the requirements need to be listed.

coolalexzb commented:
@matthewfeickert  I created a virtual python environment without any other third-part library installed before successfully, but it is weird that when I use `sphinx-build -b html source build` command, I can build the docs successfully without any notification that I need to install other libraries. 

I changed README.md to README.rst and removed m2r2 extension in `conf.py`. I think you can try to build the docs again to see if there is any other problem.

coolalexzb commented:
@matthewfeickert  One question in command line api page, the format of command lines in examples are in a mess, can you tell me how to modify it?
<img width="754" alt="截屏2020-08-19 下午5 24 26" src="https://user-images.githubusercontent.com/19755247/90696402-c8edca00-e241-11ea-8f96-62df21e4b7e2.png">

coolalexzb commented:
@matthewfeickert  The bug in CI/docs (3.8) still needs to be solved. What is the possible reason?

matthewfeickert commented:
> I created a virtual python environment without any other third-part library installed before successfully, but it is weird that when I use `sphinx-build -b html source build` command, I can build the docs successfully without any notification that I need to install other libraries.

It seems that you don't actually have a properly isolated virtual environment. If on this branch in a clean virtual environment I get the following

```
$ python -m pip install --upgrade -e .[complete]
$ python -m pip list | grep "Sphinx\|sphinx"
Sphinx                        3.2.1
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        1.0.3
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.4
$ cd docs
$ make html
Running Sphinx v3.2.1

Extension error:
Could not import extension sphinxcontrib.bibtex (exception: No module named 'sphinxcontrib.bibtex')
make: *** [Makefile:21: html] Error 2
```

> The bug in CI/docs (3.8) still needs to be solved. What is the possible reason?

If you look at the logs
```
Run python setup.py build_sphinx
  python setup.py build_sphinx
  touch docs/_build/html/.nojekyll
  shell: /bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.5/x64
running build_sphinx

Application error:
config directory doesn't contain a conf.py file (docs)
##[error]Process completed with exit code 1.
```

it tells us the same error as

```
$ python setup.py build_sphinx
running build_sphinx

Application error:
config directory doesn't contain a conf.py file (docs)
```

as `conf.py` isn't at the top level of `docs/`

```
$ find . -iname "conf.py"
./docs/source/conf.py
```
. Why are all the docs under `source/`? Is this a new recommendation from Sphinx? The first error from `make html` isn't a surprise though as you have have required the module

```
$ git grep "bibtex"
docs/source/conf.py:    "sphinxcontrib.bibtex",
```

 but haven't added it to the `docs` extra

```diff
$ git diff origin/master -- setup.py
diff --git a/setup.py b/setup.py
index c8df5b1..1d944ad 100644
--- a/setup.py
+++ b/setup.py
@@ -20,6 +20,7 @@ extras_require["develop"] = sorted(
         + ["check-manifest", "bumpversion~=0.5", "pre-commit", "twine",]
     )
 )
+extras_require["docs"] = sorted(set(["sphinx"]))
 extras_require["complete"] = sorted(set(sum(extras_require.values(), [])))
 
 setuptools.setup(extras_require=extras_require,)
```

coolalexzb commented:
@matthewfeickert  I use the same command lines as you did but I can still build the doc successfully even I denote a python version as 3.8(which is a different version with the python in my local PC) in virtual environment. As the information from your result. I just need to add to setup.py?
`extras_require["docs"] = sorted(set(["sphinxcontrib.bibtex"]))`

coolalexzb commented:
@matthewfeickert  If I don't use source file I will get an error when build docs locally. The error information is the same as CI/docs (3.8)
<img width="450" alt="截屏2020-08-19 下午11 54 37" src="https://user-images.githubusercontent.com/19755247/90718426-834bf400-e277-11ea-97cd-651a69538b3e.png">
